File: packages/php/src/Core/Criptografia.php
<?php

namespace Application\Core;

/**
 * Cifra caseira educacional.
 *
 * Estrutura do token: FBC + base64url( integridade[32] . ciphertext[n] . iv[16] )
 *
 * O keystream é gerado em blocos de 32 bytes por uma recorrência tipo-Fibonacci
 * por posição, seguida de uma rede de difusão tipo "butterfly": a cada rodada,
 * cada posição se mistura com a posição a uma distância que DOBRA (1, 2, 4, 8, 16).
 * Isso garante matematicamente que toda posição influencia toda posição do bloco
 * em exatamente log2(32) = 5 rodadas — não depende de sorte de IV.
 *
 * A quantidade de ROTAÇÃO usada dentro de cada passo (não a distância da mistura)
 * vem de uma tabela pública fixa derivada dos dígitos de Pi — só pra evitar um
 * padrão repetitivo, sem ser tratada como segredo.
 */
class Criptografia
{
    private const TAM_BLOCO = 32;

    /**
     * Distâncias de mistura geradas via Fibonacci -> N-ésimo primo, pulando
     * o único primo par (2). Como o bloco tem 32 posições (potência de 2),
     * qualquer distância ÍMPAR é automaticamente coprima com 32 - o que
     * garante matematicamente um ciclo completo de 32 posições, evitando
     * o colapso em subgrupos simétricos menores (o bug da "metade igual").
     *
     * fib(1)=1 -> 1º primo é 2 (descartado, par) -> fib(2)=1 -> idem
     * fib(3)=2 -> 2º primo=3 | fib(4)=3 -> 3º primo=5 | fib(5)=5 -> 5º primo=11
     * fib(6)=8 -> 8º primo=19 | fib(7)=13 -> 13º primo=41
     */
    private const DISTANCIAS_DIFUSAO = [3, 5, 11, 19, 41, 3, 5, 11, 19, 41]; // dobrado pra combater ataque integral

    /** Dígitos de Pi (constante pública e fixa - "nothing up my sleeve"). */
    private const DIGITOS_PI = '31415926535897932384626433832795028841971693993751058209749445923078164062862089986280348253421170679';

    public static function encrypt(string $text): string
    {
        $key = self::getKey();
        $iv = random_bytes(16);

        $encKeystream = self::gerarKeystream($key, $iv, 'enc', strlen($text));
        $ciphertext = self::xorBytes($text, $encKeystream);

        $macKey = self::gerarKeystream($key, $iv, 'mac', self::TAM_BLOCO);
        $integridade = self::checksum($iv . $ciphertext, $macKey);

        return 'FBC' . self::base64url_encode($integridade . $ciphertext . $iv);
    }

    public static function decrypt(string $text): string
    {
        if (substr($text, 0, 3) !== 'FBC') {
            throw new \Exception('Invalid text. Text must start with FBC.');
        }

        $key = self::getKey();
        $decoded = self::base64url_decode(substr($text, 3));

        $integridadeRecebida = substr($decoded, 0, self::TAM_BLOCO);
        $iv = substr($decoded, -16);
        $ciphertext = substr($decoded, self::TAM_BLOCO, -16);

        $macKey = self::gerarKeystream($key, $iv, 'mac', self::TAM_BLOCO);
        $integridadeEsperada = self::checksum($iv . $ciphertext, $macKey);

        if (!hash_equals($integridadeEsperada, $integridadeRecebida)) {
            throw new \Exception('Token adulterado ou chave incorreta.');
        }

        $encKeystream = self::gerarKeystream($key, $iv, 'enc', strlen($ciphertext));
        return self::xorBytes($ciphertext, $encKeystream);
    }

    // ---------------------------------------------------------------
    // NÚCLEO: gerador de keystream com difusão tipo butterfly
    // ---------------------------------------------------------------

    private static function gerarKeystream(string $key, string $iv, string $proposito, int $tamanho): string
    {
        $bloco = self::TAM_BLOCO;
        $ivLen = strlen($iv);
        $keyLen = strlen($key);

        $a = [];
        $b = [];
        for ($pos = 0; $pos < $bloco; $pos++) {
            $a[$pos] = ord($key[$pos % $keyLen]) ^ ord($iv[$pos % $ivLen]);
            $b[$pos] = ord($key[($pos + 1) % $keyLen]) ^ ord($iv[($pos + 1) % $ivLen]);
        }

        $saida = '';
        $roundIdx = 0;
        $posFibBase = 0;

        while (strlen($saida) < $tamanho) {
            // Uma "passada" completa de difusão: distâncias 1,2,4,8,16 —
            // garante que toda posição influenciou toda posição ao final.
            foreach (self::DISTANCIAS_DIFUSAO as $dist) {
                $novoB = [];
                for ($pos = 0; $pos < $bloco; $pos++) {
                    [$novoA, $novoBb] = self::passo(
                        $a[$pos],
                        $b[$pos],
                        $key,
                        $proposito,
                        $posFibBase + $pos,
                        $roundIdx
                    );
                    $a[$pos] = $novoA;
                    $novoB[$pos] = $novoBb;
                }

                // Combinação ASSIMÉTRICA: rotaciona só o valor próprio antes do XOR.
                // Isso é essencial - se rotacionássemos o XOR já combinado
                // (comutativo), pos e pos+16 calculariam o mesmo valor sempre
                // que $dist for exatamente metade do bloco, colapsando as duas
                // metades do bloco numa cópia idêntica uma da outra.
                $misturado = [];
                for ($pos = 0; $pos < $bloco; $pos++) {
                    $vizinho = $novoB[($pos + $dist) % $bloco];
                    $misturado[$pos] = self::rotEsquerda8($novoB[$pos], 1) ^ $vizinho;
                }
                $b = $misturado;
                $roundIdx++;
            }

            $posFibBase += $bloco;

            foreach ($b as $byte) {
                $saida .= chr($byte);
            }
        }

        return substr($saida, 0, $tamanho);
    }

    /**
     * Um passo da recorrência tipo-Fibonacci pra uma posição do bloco.
     *
     * @return array{0: int, 1: int}
     */
    private static function passo(int $a, int $b, string $key, string $proposito, int $posFib, int $roundIdx): array
    {
        $n = self::rotacaoDoRound($roundIdx);

        $soma = ($a + $b) & 0xFF;
        $soma = self::rotEsquerda8($soma, $n);
        $soma ^= ord($proposito[$posFib % strlen($proposito)]);
        // A chave agora participa de CADA rodada, não só do estado inicial.
        // Sem isso, keystream(key, iv) == keystream(key^D, iv^D16) pra
        // qualquer máscara D de 32 bytes com período 16 - o IV só deslocava
        // a chave por XOR antes da difusão começar, sem injetar entropia
        // própria no "key schedule" a cada passo.
        $soma ^= ord($key[($posFib + $roundIdx) % strlen($key)]);
        $soma = ($soma * 131) & 0xFF;

        return [$b, $soma];
    }

    /** Rotação (1 a 7) derivada dos dígitos de Pi — pública, não é segredo. */
    private static function rotacaoDoRound(int $roundIdx): int
    {
        $digitos = self::DIGITOS_PI;
        $d = (int) $digitos[$roundIdx % strlen($digitos)];
        return ($d % 7) + 1;
    }

    private static function rotEsquerda8(int $byte, int $n): int
    {
        $n &= 7;
        if ($n === 0) {
            return $byte & 0xFF;
        }
        return (($byte << $n) | ($byte >> (8 - $n))) & 0xFF;
    }

    private static function rotEsquerda32(int $val, int $n): int
    {
        $n &= 31;
        if ($n === 0) {
            return $val & 0xFFFFFFFF;
        }
        return (($val << $n) | ($val >> (32 - $n))) & 0xFFFFFFFF;
    }

    private static function xorBytes(string $dados, string $keystream): string
    {
        $out = '';
        $len = strlen($dados);
        for ($i = 0; $i < $len; $i++) {
            $out .= chr(ord($dados[$i]) ^ ord($keystream[$i]));
        }
        return $out;
    }

    /**
     * "MAC" caseiro estendido pra 32 bytes: roda um checksum de 32 bits
     * 8 vezes, mudando a semente a cada rodada, e concatena os resultados.
     */
    private static function checksum(string $dados, string $key): string
    {
        $saida = '';
        for ($rodada = 0; $rodada < 8; $rodada++) {
            $acumulador = 0x811C9DC5 ^ ($rodada * 0x01000193);
            $keyLen = strlen($key);
            $len = strlen($dados);
            for ($i = 0; $i < $len; $i++) {
                $byte = ord($dados[$i]) ^ ord($key[($i + $rodada) % $keyLen]);
                $acumulador = ($acumulador ^ $byte) & 0xFFFFFFFF;
                $acumulador = ($acumulador * 16777619) & 0xFFFFFFFF;
                $acumulador = self::rotEsquerda32($acumulador, ($i % 13) + 1);
            }
            // Finalização: sem isso, o último byte processado só passa por
            // 1 multiply+rotate antes de virar saída, e sofre avalanche
            // fraca (medimos ~25-30% em vez de ~50% nos últimos bytes).
            // Espalha o acumulador mais 3 vezes depois que TODOS os bytes
            // já entraram, garantindo que a posição de entrada deixe de
            // importar pro resultado final.
            for ($k = 0; $k < 3; $k++) {
                $acumulador ^= ($acumulador >> 16);
                $acumulador = ($acumulador * 16777619) & 0xFFFFFFFF;
                $acumulador = self::rotEsquerda32($acumulador, 13);
            }
            $saida .= chr(($acumulador >> 24) & 0xFF) . chr(($acumulador >> 16) & 0xFF)
                . chr(($acumulador >> 8) & 0xFF) . chr($acumulador & 0xFF);
        }
        return $saida; // 32 bytes
    }

    // ---------------------------------------------------------------
    // Utilitários
    // ---------------------------------------------------------------

    public static function base64url_encode(string $data): string
    {
        return rtrim(strtr(base64_encode($data), '+/', '-_'), '=');
    }

    public static function base64url_decode(string $data): string
    {
        // Rejeita qualquer caractere fora do alfabeto base64url - sem isso,
        // base64_decode() não-estrito ignora silenciosamente espaços, \n,
        // \t e caracteres inválidos, fazendo textos DIFERENTES decodificarem
        // pro mesmo token (abre espaço pra "token smuggling": um WAF/cache/log
        // vê uma string, o decrypt() vê outra).
        if ($data !== '' && !preg_match('/^[A-Za-z0-9_-]+$/', $data)) {
            throw new \Exception('Token contém caracteres inválidos.');
        }

        $padded = strtr($data, '-_', '+/');
        $mod = strlen($padded) % 4;
        if ($mod === 1) {
            throw new \Exception('Comprimento de token inválido.');
        }
        if ($mod) {
            $padded .= str_repeat('=', 4 - $mod);
        }

        $decoded = base64_decode($padded, true);
        if ($decoded === false) {
            throw new \Exception('Token base64 inválido.');
        }

        // Canonicidade: mesmo com strict=true, o último grupo de base64 pode
        // ter bits "não usados" setados como 1 em vez de 0 e ainda decodificar
        // pro mesmo byte - reencodar e comparar pega esse caso, que o
        // strict=true sozinho NÃO pega.
        if (self::base64url_encode($decoded) !== $data) {
            throw new \Exception('Token não está em forma canônica.');
        }

        return $decoded;
    }

    private static function getKey(): string
    {
        $key = getenv('FBC_KEY');
        if (strlen($key) !== 32) {
            throw new \Exception('Chave deve ter 32 bytes.');
        }
        return $key;
    }
}

File: packages/rust/src/core/criptografia.rs
//! Cifra caseira educacional — porte de `Criptografia.php`/`Criptografia.ts`.
//!
//! Estrutura do token: `FBC` + base64url( integridade[32] . ciphertext[n] . iv[16] )
//!
//! Replica a lógica byte a byte, sem "melhorias" — qualquer mudança de
//! comportamento aqui quebraria a compatibilidade com os tokens gerados pelas
//! outras linguagens.

use std::fmt;

pub const TAM_BLOCO: usize = 32;

/// Mesmas distâncias da versão PHP (Fibonacci -> N-ésimo primo, pulando o 2).
const DISTANCIAS_DIFUSAO: [usize; 10] = [3, 5, 11, 19, 41, 3, 5, 11, 19, 41];

const DIGITOS_PI: &str =
    "31415926535897932384626433832795028841971693993751058209749445923078164062862089986280348253421170679";

const ALFABETO_B64URL: &[u8; 64] =
    b"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_";

/// Erro da cifra (mensagens equivalentes às das outras linguagens).
#[derive(Debug, Clone)]
pub struct CriptoError(pub String);

impl CriptoError {
    pub fn new(mensagem: &str) -> Self {
        CriptoError(mensagem.to_string())
    }
}

impl fmt::Display for CriptoError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(&self.0)
    }
}

impl std::error::Error for CriptoError {}

fn rot_esquerda8(byte: u8, n: u32) -> u8 {
    byte.rotate_left(n & 7)
}

fn rot_esquerda32(val: u32, n: u32) -> u32 {
    val.rotate_left(n & 31)
}

fn rotacao_do_round(round_idx: usize) -> u32 {
    let digitos = DIGITOS_PI.as_bytes();
    let d = (digitos[round_idx % digitos.len()] - b'0') as u32;
    (d % 7) + 1
}

/// Um passo da recorrência tipo-Fibonacci pra uma posição do bloco.
fn passo(
    a: u8,
    b: u8,
    key: &[u8],
    proposito: &[u8],
    pos_fib: usize,
    round_idx: usize,
) -> (u8, u8) {
    let n = rotacao_do_round(round_idx);

    let mut soma = a.wrapping_add(b);
    soma = rot_esquerda8(soma, n);
    soma ^= proposito[pos_fib % proposito.len()];
    // A chave participa de CADA rodada, não só do estado inicial.
    soma ^= key[(pos_fib + round_idx) % key.len()];
    soma = soma.wrapping_mul(131);

    (b, soma) // novo a = b antigo; novo b = soma
}

/// Gera `tamanho` bytes de keystream a partir de KEY + IV + propósito.
pub fn gerar_keystream(key: &[u8], iv: &[u8], proposito: &str, tamanho: usize) -> Vec<u8> {
    let bloco = TAM_BLOCO;
    let proposito_bytes = proposito.as_bytes();

    let mut a = vec![0u8; bloco];
    let mut b = vec![0u8; bloco];
    for pos in 0..bloco {
        a[pos] = key[pos % key.len()] ^ iv[pos % iv.len()];
        b[pos] = key[(pos + 1) % key.len()] ^ iv[(pos + 1) % iv.len()];
    }

    let mut saida: Vec<u8> = Vec::new();
    let mut round_idx: usize = 0;
    let mut pos_fib_base: usize = 0;

    while saida.len() < tamanho {
        for &dist in DISTANCIAS_DIFUSAO.iter() {
            let mut novo_b = vec![0u8; bloco];
            for pos in 0..bloco {
                let (novo_a, novo_bb) = passo(
                    a[pos],
                    b[pos],
                    key,
                    proposito_bytes,
                    pos_fib_base + pos,
                    round_idx,
                );
                a[pos] = novo_a;
                novo_b[pos] = novo_bb;
            }

            // Combinação ASSIMÉTRICA: rotaciona só o valor próprio antes do XOR.
            let mut misturado = vec![0u8; bloco];
            for pos in 0..bloco {
                let vizinho = novo_b[(pos + dist) % bloco];
                misturado[pos] = rot_esquerda8(novo_b[pos], 1) ^ vizinho;
            }
            b = misturado;
            round_idx += 1;
        }

        pos_fib_base += bloco;
        saida.extend_from_slice(&b);
    }

    saida.truncate(tamanho);
    saida
}

fn xor_bytes(dados: &[u8], keystream: &[u8]) -> Vec<u8> {
    dados
        .iter()
        .zip(keystream.iter())
        .map(|(d, k)| d ^ k)
        .collect()
}

/// "MAC" caseiro: 8 rodadas de um checksum estilo FNV, concatenadas -> 32 bytes.
pub fn checksum(dados: &[u8], key: &[u8]) -> [u8; 32] {
    let mut saida = [0u8; 32];
    let key_len = key.len();

    for rodada in 0..8usize {
        let mut acumulador = 0x811c9dc5u32 ^ (rodada as u32).wrapping_mul(0x01000193);

        for (i, &byte_dado) in dados.iter().enumerate() {
            let byte = byte_dado ^ key[(i + rodada) % key_len];
            acumulador ^= byte as u32;
            acumulador = acumulador.wrapping_mul(16777619);
            acumulador = rot_esquerda32(acumulador, ((i % 13) + 1) as u32);
        }

        // Finalização: sem isso, o último byte processado sofre avalanche fraca.
        for _ in 0..3 {
            acumulador ^= acumulador >> 16;
            acumulador = acumulador.wrapping_mul(16777619);
            acumulador = rot_esquerda32(acumulador, 13);
        }

        saida[rodada * 4] = (acumulador >> 24) as u8;
        saida[rodada * 4 + 1] = (acumulador >> 16) as u8;
        saida[rodada * 4 + 2] = (acumulador >> 8) as u8;
        saida[rodada * 4 + 3] = acumulador as u8;
    }

    saida
}

fn valor_b64(c: u8) -> Option<u8> {
    match c {
        b'A'..=b'Z' => Some(c - b'A'),
        b'a'..=b'z' => Some(c - b'a' + 26),
        b'0'..=b'9' => Some(c - b'0' + 52),
        b'-' => Some(62),
        b'_' => Some(63),
        _ => None,
    }
}

pub fn base64url_encode(data: &[u8]) -> String {
    let mut out = String::new();
    for chunk in data.chunks(3) {
        let b0 = chunk[0] as u32;
        let b1 = if chunk.len() > 1 { chunk[1] as u32 } else { 0 };
        let b2 = if chunk.len() > 2 { chunk[2] as u32 } else { 0 };
        let n = (b0 << 16) | (b1 << 8) | b2;

        out.push(ALFABETO_B64URL[((n >> 18) & 63) as usize] as char);
        out.push(ALFABETO_B64URL[((n >> 12) & 63) as usize] as char);
        if chunk.len() > 1 {
            out.push(ALFABETO_B64URL[((n >> 6) & 63) as usize] as char);
        }
        if chunk.len() > 2 {
            out.push(ALFABETO_B64URL[(n & 63) as usize] as char);
        }
    }
    out
}

pub fn base64url_decode(data: &str) -> Result<Vec<u8>, CriptoError> {
    if !data.is_empty() && !data.bytes().all(|c| valor_b64(c).is_some()) {
        return Err(CriptoError::new("Token contém caracteres inválidos."));
    }

    if data.len() % 4 == 1 {
        return Err(CriptoError::new("Comprimento de token inválido."));
    }

    let bytes = data.as_bytes();
    let mut out: Vec<u8> = Vec::new();
    let mut i = 0;
    while i + 4 <= bytes.len() {
        let n = ((valor_b64(bytes[i]).unwrap() as u32) << 18)
            | ((valor_b64(bytes[i + 1]).unwrap() as u32) << 12)
            | ((valor_b64(bytes[i + 2]).unwrap() as u32) << 6)
            | (valor_b64(bytes[i + 3]).unwrap() as u32);
        out.push((n >> 16) as u8);
        out.push((n >> 8) as u8);
        out.push(n as u8);
        i += 4;
    }

    let restante = bytes.len() - i;
    if restante == 2 {
        let n = ((valor_b64(bytes[i]).unwrap() as u32) << 18)
            | ((valor_b64(bytes[i + 1]).unwrap() as u32) << 12);
        out.push((n >> 16) as u8);
    } else if restante == 3 {
        let n = ((valor_b64(bytes[i]).unwrap() as u32) << 18)
            | ((valor_b64(bytes[i + 1]).unwrap() as u32) << 12)
            | ((valor_b64(bytes[i + 2]).unwrap() as u32) << 6);
        out.push((n >> 16) as u8);
        out.push((n >> 8) as u8);
    }

    // Canonicidade: reencoda e compara — pega bits não-canônicos no último grupo.
    if base64url_encode(&out) != data {
        return Err(CriptoError::new("Token não está em forma canônica."));
    }

    Ok(out)
}

pub fn get_key() -> Result<Vec<u8>, CriptoError> {
    let key = std::env::var("FBC_KEY").unwrap_or_default();
    if key.len() != 32 {
        return Err(CriptoError::new("Chave deve ter 32 bytes."));
    }
    Ok(key.into_bytes())
}

/// N bytes aleatórios (equivalente ao `random_bytes` do PHP/Node).
pub fn random_bytes(n: usize) -> Vec<u8> {
    use std::io::Read;

    if let Ok(mut f) = std::fs::File::open("/dev/urandom") {
        let mut buf = vec![0u8; n];
        if f.read_exact(&mut buf).is_ok() {
            return buf;
        }
    }

    // Fallback: xorshift semeado com tempo + pid.
    let mut estado = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.as_nanos() as u64)
        .unwrap_or(0x9E37_79B9_7F4A_7C15);
    estado ^= std::process::id() as u64;

    let mut out = Vec::with_capacity(n);
    for _ in 0..n {
        estado ^= estado << 13;
        estado ^= estado >> 7;
        estado ^= estado << 17;
        out.push((estado & 0xff) as u8);
    }
    out
}

/// `encrypt` a partir de bytes arbitrários (equivalente ao `encrypt(Buffer)`).
pub fn encrypt_bytes(texto: &[u8]) -> Result<String, CriptoError> {
    let key = get_key()?;
    let iv = random_bytes(16);

    let enc_keystream = gerar_keystream(&key, &iv, "enc", texto.len());
    let ciphertext = xor_bytes(texto, &enc_keystream);

    let mac_key = gerar_keystream(&key, &iv, "mac", TAM_BLOCO);
    let mut dados_mac = iv.clone();
    dados_mac.extend_from_slice(&ciphertext);
    let integridade = checksum(&dados_mac, &mac_key);

    let mut corpo = Vec::with_capacity(32 + ciphertext.len() + 16);
    corpo.extend_from_slice(&integridade);
    corpo.extend_from_slice(&ciphertext);
    corpo.extend_from_slice(&iv);

    Ok(format!("FBC{}", base64url_encode(&corpo)))
}

/// `encrypt` de uma string UTF-8.
pub fn encrypt(texto: &str) -> Result<String, CriptoError> {
    encrypt_bytes(texto.as_bytes())
}

/// Comparação em tempo constante (equivalente ao `timingSafeEqual`).
fn iguais_tempo_constante(a: &[u8], b: &[u8]) -> bool {
    if a.len() != b.len() {
        return false;
    }
    let mut diff = 0u8;
    for i in 0..a.len() {
        diff |= a[i] ^ b[i];
    }
    diff == 0
}

pub fn decrypt(token: &str) -> Result<String, CriptoError> {
    if token.as_bytes().get(0..3) != Some(b"FBC") {
        return Err(CriptoError::new("Invalid text. Text must start with FBC."));
    }

    let key = get_key()?;
    let decoded = base64url_decode(&token[3..])?;

    if decoded.len() < TAM_BLOCO + 16 {
        return Err(CriptoError::new("Token adulterado ou chave incorreta."));
    }

    let integridade_recebida = &decoded[0..TAM_BLOCO];
    let iv = &decoded[decoded.len() - 16..];
    let ciphertext = &decoded[TAM_BLOCO..decoded.len() - 16];

    let mac_key = gerar_keystream(&key, iv, "mac", TAM_BLOCO);
    let mut dados_mac = iv.to_vec();
    dados_mac.extend_from_slice(ciphertext);
    let integridade_esperada = checksum(&dados_mac, &mac_key);

    if !iguais_tempo_constante(&integridade_esperada, integridade_recebida) {
        return Err(CriptoError::new("Token adulterado ou chave incorreta."));
    }

    let enc_keystream = gerar_keystream(&key, iv, "enc", ciphertext.len());
    let plaintext = xor_bytes(ciphertext, &enc_keystream);
    Ok(String::from_utf8_lossy(&plaintext).into_owned())
}
